ops/main.py

```diff
@@ -30,24 +30,43 @@ def _get_charm_dir():
 def _load_metadata(charm_dir):
     with open(charm_dir / 'metadata.yaml') as f:
         metadata = yaml.load(f, Loader=yaml.SafeLoader)
+
+    functions_meta = charm_dir / 'functions.yaml'
+    actions_meta = charm_dir / 'actions.yaml'
+    if functions_meta.exists():
+        # TODO: add a version check and a warning here to make sure functions are not skipped without notice pre-2.7.0 Juju versions.
+        if actions_meta.exists():
+            raise RuntimeError('charm must not mix functions and actions metadata')
+        metadata['functions_type'] = 'functions'
+        metadata['functions'] = yaml.safe_load(functions_meta.read_text())
+    elif actions_meta.exists():
+        metadata['functions_type'] = 'actions'
+        metadata['functions'] = yaml.safe_load(actions_meta.read_text())
```

Comment:
We should try to avoid that kind of mixing of data that comes from a user file with data that we inject because we were lazy and didn't want to change the code to pass data around properly. Imagine what happens now if neither functions.yaml or actions.yaml exists, but the user puts a key labeled "functions" in metadata.yaml. Oops. It also reads improperly/confusingly, per earlier not in the PR.. it's the "raw data", except for things we stashed on it for convenience.